samples/cxl_host_exerciser/he_cache_test.h

```diff
@@ -0,0 +1,830 @@
+
+    command::ptr_t test(nullptr);
+    CLI::App *app = nullptr;
+    for (auto kv : commands_) {
+      if (*kv.first) {
+        app = kv.first;
+        test = kv.second;
+        break;
+      }
+    }
+    if (!test) {
+      std::cerr << "no command specified\n";
+      return exit_codes::not_run;
+    }
+
+    auto console_sink = std::make_shared<spdlog::sinks::stdout_color_sink_mt>();
+    logger_ = std::make_shared<spdlog::logger>(test->name(), console_sink);
+    spdlog::register_logger(logger_);
+    logger_->set_level(spdlog::level::from_str(log_level_));
+    current_command_ = test;
+    if (find_dev_feature() != 0) {
+      cerr << "fails to find feature" << endl;
+      return exit_codes::exception;
+    };
+
+    int res = open_handle();
+    if (res != exit_codes::not_run) {
+      return res;
+    }
+
+    return run(app, test);
+  }
+
+  virtual int run(CLI::App *app, command::ptr_t test) {
+    int res = exit_codes::not_run;
+    current_command_ = test;
+
+    try {
+      std::future<int> f = std::async(std::launch::async, [this, test, app]() {
+        return test->run(this, app);
+      });
+      auto status = f.wait_for(std::chrono::milliseconds(timeout_msec_));
+      if (status == std::future_status::timeout) {
+        std::cerr << "Error: test timed out" << std::endl;
+        current_command_->stop();
+        throw std::runtime_error("timeout");
+      }
+      res = f.get();
+    } catch (std::exception &ex) {
+      res = exit_codes::exception;
+    }
+
+    current_command_.reset();
+    return res;
+  }
+
+  template <class T> CLI::App *register_command() {
+    command::ptr_t cmd(new T());
+    auto sub = app_.add_subcommand(cmd->name(), cmd->description());
+    cmd->add_options(sub);
+    commands_[sub] = cmd;
+    return sub;
+  }
+
+  uint64_t read64(uint32_t offset) {
+    uint64_t value = *((uint64_t *)(mmio_base_ + offset));
+    return value;
+  }
+
+  void write64(uint32_t offset, uint64_t value) {
+    *((uint64_t *)(mmio_base_ + offset)) = value;
+    return;
+  }
+
+  uint32_t read32(uint32_t offset) {
+    uint32_t value = *((uint64_t *)(mmio_base_ + offset));
+    return value;
+  }
+
+  void write32(uint32_t offset, uint32_t value) {
+    *((uint32_t *)(mmio_base_ + offset)) = value;
+    return;
+  }
+
+  command::ptr_t current_command() const { return current_command_; }
+
+  bool allocate_dsm(size_t len = KiB(4), uint32_t numa_node = 0) {
+
+    int res = 0;
+    void *ptr = NULL;
+    struct dfl_cxl_cache_dma_map dma_map;
+
+    memset(&dma_map, 0, sizeof(dma_map));
+    ptr = alloc_4kb_hugepage();
+    if (!ptr) {
+      cerr << "Fails to allocate 4k huge page:" << strerror(errno) << endl;
+      return false;
+    }
+
+    cout << "DSM buffer numa node: " << numa_node << endl;
+    dma_map.argsz = sizeof(dma_map);
+    dma_map.user_addr = (__u64)ptr;
+    dma_map.length = len;
+    dma_map.numa_node = numa_node;
+    dma_map.csr_array[0] = DFL_CXL_CACHE_DSM_BASE; // 0x030
```

Comment:
Copying the address here in the comment defeats the purpose of using the \#define.  Please remove.